File: swift/BCEnvelope/Tests/BCEnvelopeTests/EncapsulationTests.swift
import Testing
import BCComponents
import BCEnvelope
import Foundation

struct EncapsulationTests {
    private func testScheme(_ scheme: EncapsulationScheme) throws {
        let (privateKey, publicKey) = scheme.keypair()
        let envelope = Envelope(plaintextHello)

        let encryptedEnvelope = try envelope
            .encryptToRecipient(publicKey)
            .checkEncoding()

        let decryptedEnvelope = try encryptedEnvelope.decryptToRecipient(privateKey)
        #expect(envelope.structuralDigest == decryptedEnvelope.structuralDigest)
    }

    @Test func testEncapsulation() throws {
        try testScheme(.x25519)
        try testScheme(.mlkem512)
        try testScheme(.mlkem768)
        try testScheme(.mlkem1024)
    }
}

File: csharp/BCEnvelope/BCEnvelope.Tests/EncapsulationTests.cs
using BlockchainCommons.BCComponents;
using BlockchainCommons.BCEnvelope;

namespace BlockchainCommons.BCEnvelope.Tests;

public sealed class EncapsulationTests
{
    private static void TestScheme(EncapsulationScheme scheme)
    {
        var (privateKey, publicKey) = scheme.Keypair();
        var envelope = TestData.HelloEnvelope();
        var encryptedEnvelope = envelope
            .EncryptToRecipient(publicKey)
            .CheckEncoding();
        var decryptedEnvelope = encryptedEnvelope
            .DecryptToRecipient(privateKey);
        Assert.Equal(
            envelope.StructuralDigest(),
            decryptedEnvelope.StructuralDigest());
    }

    [Fact]
    public void TestEncapsulation()
    {
        TagsRegistry.RegisterTags();

        TestScheme(EncapsulationScheme.X25519);
        TestScheme(EncapsulationScheme.MLKEM512);
        TestScheme(EncapsulationScheme.MLKEM768);
        TestScheme(EncapsulationScheme.MLKEM1024);
    }
}
